ui/raidboss/emulator/data/PopupTextAnalysis.js

```diff
@@ -60,132 +116,121 @@ export default class PopupTextAnalysis extends StubbedPopupText {
+        const ret = await this.triggerResolvers[i].IsResolved(logObj);
+        if (ret)
+          this.triggerResolvers[i] = false;
+      }
+
+      this.triggerResolvers = this.triggerResolvers.filter((resolver) => resolver);
     }
   }
 
   _onTriggerInternalCheckSuppressed(trigger, when) {
-    this.currentFunction = '_onTriggerInternalCheckSuppressed';
     const ret = super._onTriggerInternalCheckSuppressed(trigger, when);
     this.currentTriggerStatus.suppressed = ret;
-    if (ret) {
-      this.delayResolver();
-      this.promiseResolver();
-      this.runResolver({ trigger: trigger });
-    }
     return ret;
   }
 
   _onTriggerInternalCondition(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalCondition';
+    this.currentTriggerStatus = triggerHelper.resolver.status;
     const ret = super._onTriggerInternalCondition(triggerHelper);
     this.currentTriggerStatus.condition = ret;
-    if (!ret) {
-      this.delayResolver();
-      this.promiseResolver();
-      this.runResolver(triggerHelper);
-    }
     return ret;
   }
 
-  // This is pointless right now since _onTriggerInternalHelperDefaults won't ever
-  // modify the data state, but adding it for consistency
-  _onTriggerInternalHelperDefaults(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalHelperDefaults';
-    super._onTriggerInternalHelperDefaults(triggerHelper);
-  }
-
   _onTriggerInternalPreRun(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalPreRun';
+    this.currentTriggerStatus = triggerHelper.resolver.status;
     super._onTriggerInternalPreRun(triggerHelper);
   }
 
   _onTriggerInternalDelaySeconds(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalDelaySeconds';
+    this.currentTriggerStatus = triggerHelper.resolver.status;
     // Can't inherit the default logic for delay since we don't
     // want to delay for mass processing of the timeline
     const delay = 'delaySeconds' in triggerHelper.trigger ? triggerHelper.valueOrFunction(triggerHelper.trigger.delaySeconds) : 0;
     this.currentTriggerStatus.delay = delay;
-    return new Promise((res) => {
-      this.delayResolver();
-      res();
-    });
+    if (!delay || delay <= 0)
+      return null;
+    return triggerHelper.resolver.setDelay(triggerHelper.now + (delay * 1000));
   }
 
   _onTriggerInternalDurationSeconds(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalDurationSeconds';
+    this.currentTriggerStatus = triggerHelper.resolver.status;
     super._onTriggerInternalDurationSeconds(triggerHelper);
   }
 
   _onTriggerInternalSuppressSeconds(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalSuppressSeconds';
+    this.currentTriggerStatus = triggerHelper.resolver.status;
     super._onTriggerInternalSuppressSeconds(triggerHelper);
   }
 
   _onTriggerInternalPromise(triggerHelper) {
-    this.currentFunction = '_onTriggerInternalPromise';
+    this.currentTriggerStatus = triggerHelper.resolver.status;
```

Comment:
I'm a little confused what `this.currentTriggerStatus` is doing here, and elsewhere.  You set `this.currentTriggerStatus` in a lot of places, but in reading through the code it seems like most of those could be dropped and instead `triggerHelper.resolver.status.promise = ret` could be used instead.  What's the side effect of setting this that I don't understand?

Reply:
See my comment on the review above this, https://github.com/quisquous/cactbot/pull/2862#discussion_r632994855

Reply:
This doesn't really answer my question here about storing a handle to `triggerHelper.resolver.status` in `this.currentTriggerStatus`.  Where is `this.currentTriggerStatus` being used in a place that you don't have the `triggerHelper` or the `resolver`?

Reply:
These functions in `PopupTextAnalysis`:

- _onTriggerInternalCheckSuppressed
- _makeTextElement
- _createTextFor
- _playAudioFile

Reply:
What about: `_makeTextElement`, `_createTextFor`, and `_playAudioFile` get a `triggerHelper` parameter on their functions.  It looks pretty straightforward to pass that through.

`_onTriggerInternalCheckSuppressed` is weirder, but I'm not sure why status needs to store suppressed, as it doesn't seem to get used anywhere? Could it just not store that at all and suppressed triggers can just be dropped entirely?

Maybe if the right information can get passed in here, then the currentTriggerStatus setting can go away and it can all be done through triggerHelper.

Reply:
> What about: `_makeTextElement`, `_createTextFor`, and `_playAudioFile` get a `triggerHelper` parameter on their functions. It looks pretty straightforward to pass that through.
This would resolve the issue in general, yes. Should I include that as part of this PR, given that it's going to require changes in both raidboss and raidemulator even if it's just that change?

> `_onTriggerInternalCheckSuppressed` is weirder, but I'm not sure why status needs to store suppressed, as it doesn't seem to get used anywhere? Could it just not store that at all and suppressed triggers can just be dropped entirely?
I decided to store the suppressed state in case we want to indicate it later. As it is right now, there's no visual indicator of suppressed triggers, however it's still visible that a trigger is suppressed in the JSON output due to being stored here.

> Maybe if the right information can get passed in here, then the currentTriggerStatus setting can go away and it can all be done through triggerHelper.
These changes would allow getting rid of the `this.currentTriggerStatus` property, however the circular reference of `triggerHelper.resolver.triggerHelper` would still be needed because the callback needs `triggerHelper` which isn't available otherwise, and `triggerHelper.resolver` is needed elsewhere as well.

Reply:
> > What about: `_makeTextElement`, `_createTextFor`, and `_playAudioFile` get a `triggerHelper` parameter on their functions. It looks pretty straightforward to pass that through.
> 
> This would resolve the issue in general, yes. Should I include that as part of this PR, given that it's going to require changes in both raidboss and raidemulator even if it's just that change?

Yeah, I think it would be fine to include it here.  You could also submit it as a separate PR to land first if you wanted to separate it out.  Those changes seem like they should be pretty minor plumbing.

> > `_onTriggerInternalCheckSuppressed` is weirder, but I'm not sure why status needs to store suppressed, as it doesn't seem to get used anywhere? Could it just not store that at all and suppressed triggers can just be dropped entirely?
> 
> I decided to store the suppressed state in case we want to indicate it later. As it is right now, there's no visual indicator of suppressed triggers, however it's still visible that a trigger is suppressed in the JSON output due to being stored here.

Personally I think it would be better to remove the noise of suppressed triggers.  I have really never had a situation in which it was useful to look at them.  Triggers also get suppressed prior to the point of constructing a triggerHelper too.

> > Maybe if the right information can get passed in here, then the currentTriggerStatus setting can go away and it can all be done through triggerHelper.
> 
> These changes would allow getting rid of the `this.currentTriggerStatus` property, however the circular reference of `triggerHelper.resolver.triggerHelper` would still be needed because the callback needs `triggerHelper` which isn't available otherwise, and `triggerHelper.resolver` is needed elsewhere as well.

I think it might be possible to remove the resolver -> triggerHelper lookup (maybe by setting final somewhere else?) but I think it's not a big deal and this could land with that circular reference.  Those two objects are tightly coupled together.

It's really `this.currentTriggerStatus` as global state being set everywhere that really feels like the blocking issue to me.

Reply:
Do you have a preference on if triggerHelper should go first or last in the arguments?

Reply:
I don't feel strongly.  It looks like there's only one function (`_addTextFor`) that takes both a `triggerHelper` and something else so far.  I can see both "always go first" or "always goes last" making sense.

Reply:
I opted to prepend the parameter since it's the more common parameter in the class scope. Changes are pushed, please let me know if there are any other adjustments for this.